Add option to enable worker graceful shutdown

sdaberdaku opened the pull request:
In this PR I have automated the process of enabling the Worker [Graceful Shutdown](https://trino.io/docs/current/admin/graceful-shutdown.html) feature.

Enabling the option will:
 * Add a `preStop` lifecycle event to all worker Pods;
 * Set the `shutdown.grace-period` configuration property to `gracePeriod`;
 * Configure the workers' `accessControl` since the `default` system access control [does not allow graceful shutdowns](https://trino.io/docs/current/admin/graceful-shutdown.html).

I also added a test to check if the graceful shutdown endpoint can be successfully called on the worker Pods.

Fixes #189

nineinchnick commented on charts/trino/templates/configmap-worker.yaml:
```diff
@@ -57,10 +57,19 @@ data:
     {{- end }}
 
+{{- if .Values.worker.gracefulShutdown.enabled }}
+  access-control.properties: |
```
I'm not entirely sure graceful shutdown is the only reason to configure access control on workers. #213 has a more generic approach.

nineinchnick replied:
See my other comment about failing if both a custom preStop hook is configured, and graceful shutdown is enabled. Maybe the same can be applied here - allow configuring custom access control only if graceful shutdown is not enabled.

sdaberdaku replied:
I have checked the documentation and graceful shutdown is the only place where authorization is mentioned on worker nodes. And basically you can either [set system access control to `allow-all`](https://trino.io/docs/current/admin/graceful-shutdown.html), or to setup [file access control with write access to the `system_information` schema](https://trino.io/docs/current/security/file-system-access-control.html#system-information-rules). I went with the second because I thought a least permissions approach was better. 

In any case, I think I can use the multiple authorization configurations feature and allow other authorization mechanisms alongside this.

nineinchnick commented on charts/trino/templates/tests/test-graceful-shutdown.yaml:
```diff
@@ -0,0 +1,51 @@
+      command: ["sh", "-c"]
+      args:
+        - >-
+          curl -v -X PUT
```
Shouldn't this kill the worker pods, to verify the `preStop` lifecycle hook? I also don't see any assertion - ideally this should start a long-running query (a select query on the blackhole connector with `page_processing_delay` set) and verify it was able to finish before the cluster was restarted - verify the uptime.

sdaberdaku replied:
The ` --fail-with-body` will return non-zero exit code if the `curl` call returns any HTTP code >= 400.
So basically I am testing that the graceful shutdown authorization was configured correctly.
But I get your point, I think I can write a better test, maybe using a kubectl image and actually killing the worker pod.

nineinchnick commented on charts/trino/values.yaml:
```diff
@@ -649,21 +651,43 @@ worker:
+  #  2) Set the `shutdown.grace-period` configuration property to `gracePeriod`;
+  #  3) Configure the workers' `accessControl` since the `default` system access control [does not allow graceful
+  #    shutdowns](https://trino.io/docs/current/admin/graceful-shutdown.html).
+  # The user must set the `terminationGracePeriodSeconds` to a value of at least two times the configured `gracePeriod`.
```
I don't like that we can't make this feature complete. It's so easy to miss this note, and we can't really verify that anywhere in the chart - we can't parse intervals to compare them.

We should remove the default, set it based on the grace period (can it parse the same units?), and allow users to override it, same as with the worker `preStop` lifecycle policy.

sdaberdaku replied:
I think we can require the user to provide the two values in seconds (and add the unit suffix when required), and do a sanity check on them with a `fail` if incompatible.

nineinchnick commented on charts/trino/values.yaml:
````diff
@@ -649,21 +651,43 @@ worker:
-  #      command: ["/bin/sh", "-c", "curl -v -X PUT -d '\"SHUTTING_DOWN\"' -H \"Content-type: application/json\" http://localhost:8081/v1/info/state"]
+  #      command: ["/bin/sh", "-c", "sleep 120"]
+  # ```
+  # If provided, it will override the `preStop` lifecycle event configured by `gracefulShutdown`.
````
This is important, it should be mentioned before the example.

nineinchnick commented on charts/trino/values.yaml:
```diff
@@ -649,21 +651,43 @@ worker:
+    gracePeriod: 2m
+    accessControl:
+      user: admin
+      configFile: graceful-shutdown-rules.json
```
Why this is configurable?

nineinchnick commented on charts/trino/templates/deployment-worker.yaml:
```diff
@@ -144,7 +156,21 @@ spec:
             successThreshold: {{ .Values.worker.readinessProbe.successThreshold | default 1 }}
           lifecycle:
+            {{- if .Values.worker.lifecycle }}
             {{- toYaml .Values.worker.lifecycle | nindent 12 }}
```
I wonder if we should explicitly call `fail` here if `.Values.worker.gracefulShutdown.enabled` is true. If a custom lifecycle policy is used here, all the graceful shutdown configuration can also be applied manually.

sdaberdaku replied:
agreed

nineinchnick reviewed (commented):
This looks good, but I have some comments about making it even more robust, and not allowing users to make configuration mistakes.

sdaberdaku commented:
Hello @nineinchnick,

Thank you so much for finding the time to review this PR!
I will try to implement all the required changes, hopefully during this week and I will let you know then they are ready.

nineinchnick commented:
Thanks! No rush and we can always chat on the Trino slack if anything is not clear enough

sdaberdaku commented:
Dear @nineinchnick,

I have addressed all your comments.

Regarding the test implementation, I considered using a long-running query; however, I was uncertain about its appropriateness as it seemed more like a Trino test than a Helm Chart test. Ultimately, my testing approach involves deleting a worker pod using `kubectl` and verifying that the logs of the affected pod contain the "Shutdown requested" string. This method effectively tests both the worker's authorization and the lifecycle configuration.

Looking forward for your feedback!

nineinchnick commented on charts/trino/templates/deployment-worker.yaml:
```diff
@@ -150,7 +166,24 @@ spec:
           lifecycle:
+            {{- if .Values.worker.lifecycle }}
+            {{- if .Values.worker.gracefulShutdown.enabled }}
+            {{- fail "The `worker.lifecycle` configuration conflicts with `worker.gracefulShutdown`. Either disable `worker.gracefulShutdown` (and apply the related configurations manually), or remove `worker.lifecycle`." }}
```
Small nitpick, avoid parentheses, ref: https://developers.google.com/style/parentheses
```suggestion
            {{- fail "The `worker.lifecycle` configuration conflicts with `worker.gracefulShutdown`. Either disable `worker.gracefulShutdown`, and apply the related configurations manually, or remove `worker.lifecycle`." }}
```

sdaberdaku replied:
done!

nineinchnick reviewed (approved):
This is looking great now, thanks! Resolve the conflict and we can merge this.

sdaberdaku commented:
I rebased and addressed your last comment. Should be ready now!

nineinchnick commented on charts/trino/values.yaml:
```diff
@@ -661,20 +663,42 @@ worker:
+  gracefulShutdown:
+    enabled: false
+    gracePeriodSeconds: 120
+  # worker.gracefulShutdown -- Configure [graceful
+  # shutdown](https://trino.io/docs/current/admin/graceful-shutdown.html)
+  #
+  # Enabling this feature will:
+  #  1) Add a `preStop` lifecycle event to all worker Pods;
+  #  2) Set the `shutdown.grace-period` configuration property to `gracePeriod`;
+  #  3) Configure the workers' `accessControl` since the `default` system access control [does not allow graceful
+  #    shutdowns](https://trino.io/docs/current/admin/graceful-shutdown.html).
+  # The user must set the `terminationGracePeriodSeconds` to a value of at least two times the configured `gracePeriodSeconds`.
+  # The worker that receives the graceful shutdown request [will sleep for `gracePeriod` twice](https://trino.io/docs/current/admin/graceful-shutdown.html#shutdown-behavior).
+  #
+  # Enabling `worker.gracefulShutdown` conflicts with `worker.lifecycle`. If you need to provide a custom
+  # `worker.lifecycle` configuration, and you want to enable `worker.gracefulShutdown`, you have to do so manually.
```
Newlines don't get preserved when this is rendered in the chart readme. I think this is too much detail anyway, so maybe remove it. I reworded other parts too:
```suggestion
  # worker.gracefulShutdown -- Configure [graceful
  # shutdown](https://trino.io/docs/current/admin/graceful-shutdown.html)
  # in order to ensure that workers terminate without affecting running queries,
  # given a sufficient grace period.
  # When enabled, the value of `worker.terminationGracePeriodSeconds` must be at least two times greater than the configured `gracePeriodSeconds`.
  # Enabling `worker.gracefulShutdown` conflicts with `worker.lifecycle`. When a custom
  # `worker.lifecycle` configuration needs to be used, graceful shutdown must be configured manually.
```

sdaberdaku replied:
done!

nineinchnick reviewed (approved):
One last nitpick to get better docs.
